### ISSUE TITLE
Sort out completion auto-open and show options

### ISSUE THREAD

**Issue by mlochbaum** [The-Compiler/qutebrowser#1763]
So the recent completion changes have revealed the mess that is deciding when to show the completion window. I understand that no one has any clue how this should be configured, so I think it is best to start with what different people want to do and go from there.

Here are three patterns that should be supported:
* Open whenever there is a completion available (default)
* Open only once the user asks for a completion with up, down, or tab (my preference)
* Never open. (Does anyone use this?)

If these are the only possibilities I would be inclined to have a single option `show` with possible values `always`, `auto`, and `never`. Are there other ways that people want to use the completion window?

**Comment by The-Compiler** [comment]
That proposal sounds good to me. I'm just not sure if we'll be able to auto-migrate current configs with the current code to that, but if not, I guess that's fine too.

**Comment by rcorre** [comment]
> I'm just not sure if we'll be able to auto-migrate current configs

| show  | auto-open  | new setting |
|-------|------------|-------------|
| false | don't care | never       |
| true  | false      | auto        |
| true  | true       | always      |

**Comment by The-Compiler** [comment]
:laughing: Sure, but I mean with the current config migration possibilities in `config.py`.

**Comment by mlochbaum** [comment]
It's probably enough to change `show = false` to `show = never` and `show = true` to `show = always`, and drop `auto-open`. It breaks the `auto` case, but that's already broken right now.

**Comment by rcorre** [comment]
Is it? `auto-open=false` seems to do what I'd expect right now (don't show completion until I hit TAB). So ideally users with `auto-open=false` would have `show=auto` using the new settings. But I guess what @The-Compiler is saying is that our current config updating code isn't set up to handle translating two config settings into one.

**Comment by rcorre** [comment]
On second thought, you're right. Considering the current config is already pretty confusing, its probably not worth adding confusing code to translate old configs.

**Comment by The-Compiler** [comment]
There is already the possibility of doing [simple migrations](https://github.com/The-Compiler/qutebrowser/blob/v0.8.2/qutebrowser/config/config.py#L296-L303), just not from two options into one (because then the config would need to be parsed in two passes).

What @mlochbaum says sounds good (via `CHANGED_OPTIONS`), though I also think `auto-open` is working like it should right now, no?

**Comment by mlochbaum** [comment]
Setting `auto-open=false` and `show=true` currently does the same thing for me as `show=false`, that is, the completion widget never opens. This is true even if I restart the browser.

**Comment by The-Compiler** [comment]
@mlochbaum Even if you hit tab?

**Comment by mlochbaum** [comment]
Now it does. Interesting story:
* Initially I had `auto-open=true` and `show=true`. Completions opened automatically and worked.
* I set `auto-open` to `false` on the settings page and completion stopped working entirely. No widget and no completion. This persisted through `:restart`.
* I went through several configurations with one or both show settings false, each time checking completions, restarting, and checking again. At no point did completions or the widget appear.
* On trying to change both settings to `true` through the settings page, I found that the settings were not in fact changing, even though Javascript was enabled. So I set the value with `:set`. Everything worked as it had initially.
* I then set `auto-open` to `false` with `:set`. Now it works as expected: the widget is hidden until I press up, down, or tab, and shown when I do.

Needless to say, I find this somewhat confusing.

I noticed that the completion window will auto-open when pressing `:` before I ask for a completion. This seems wrong to me.

**Comment by The-Compiler** [comment]
> * I set `auto-open` to `false` on the settings page and completion stopped working entirely. No widget and no completion. This persisted through `:restart`.

That's odd.

> * On trying to change both settings to `true` through the settings page, I found that the settings were not in fact changing, even though Javascript was enabled.

Unfortunately you're not the only one with random trouble on the settings page :unamused: 

> I noticed that the completion window will auto-open when pressing `:` before I ask for a completion. This seems wrong to me.

You mean with `auto-open = false`?

**Comment by mlochbaum** [comment]
Okay, new information that sort of explains some of that. If qutebrowser is opened with `auto-open=false` and `show=true`, then completions will not appear (neither completions nor the widget) until `auto-open` is set to true. If it's then set back to false, things work like they should.

The `:` bug is with `auto-open=false`. It appears it only happens if the completion window was used (not just opened--see the next paragraph) in the previous command. This doesn't happen for other commands, like the keybinding for `bookmark-load`.

Another bug: if `auto-open` is false, then pressing up, down, or tab causes the completion window to open, but doesn't complete the current command. Subsequently pressing one of these completes as usual. This is particularly jarring if `show=false`, since it will work just like `auto-open=true` except that the first tab, up, or down pressed for a given command is ignored (except for the `:` bug above).

**Comment by mlochbaum** [comment]
@rcorre Are you planning to work on this? If not, I'll pick it up.

**Comment by rcorre** [comment]
@mlochbaum go for it. Let me know if you have any questions as I've been working with the completion code a lot recently.

**Comment by mlochbaum** [comment]
It seems like the best way forward is just to remove all of the `auto-open` code from `Completer` and move it to `CompletionView` with `show`. This gets rid of the whole `_open_completion_if_needed` function, replacing it with code that sets `enabled` appropriately when `completion_item_focus` is called. It also means that the completion model will be connected even when `auto-open` is false, but I don't think we ever wanted to suppress that anyway.

Does this sound right? I also wanted to check whether adding `on_clear_completion_selection` is the best way to turn `enabled` off again at the end of completion if `auto-open` is false.

**Comment by rcorre** [comment]
>  It seems like the best way forward is just to remove all of the auto-open code from Completer and move it to CompletionView with show

Yes, I'd actually been meaning to do this. `Completer` could unconditionally call `show` without checking `enabled`, and `CompletionView` would override `show` and only pass it through if it is enabled.

> the completion model will be connected even when auto-open is false

I think that's fine. It might cause TAB to cycle through completions without showing the completion menu when `show=never`, which might even be preferable.

> turn enabled off again at the end of completion if auto-open is false

I'm actually not sure how this works right now. I don't think auto-open works by
toggling `enabled` though. I think it works by calling `hide` if it comes up
with `None` as the completion model for the current input.

Come to think of it, I don't think you need to mess with much logic here, its
just a matter of changing the configs that get checked. That is, check for
`show=auto` instead of `auto-open=false` and `show=always` instead of
`auto-open=always`.

**Comment by mlochbaum** [comment]
I'm trying to fix all of the broken behavior described above rather than just change the option system. Maybe I should have split this into a new issue.

Removing `enabled` and just calling `hide` and `show` directly mostly works, but the fact that `_update_completion` is called on every keypress means that the completion will get hidden too often when `auto-open` is false (for instance, when pressing backspace during a completion). Is there a good way to trigger an initial (conditional) show only at the start of a completion?

**Comment by mlochbaum** [comment]
So there are actually two different behaviors that `auto-open` could support. The command window can be re-hidden either when the current completion is done or when the entire command is finished. I think the old behavior was to reset after the command, but I kind of like the other version better since it cuts down on the number of instances when the completion window is open and I didn't ask for it. It's possible to do both, although I don't know what to call them to make the distinction clear.

**Comment by rcorre** [comment]
Is there currently broken behavior? I thought your problem turned out to be something weird with the config page, auto-open works as expected for me.

As far as the behaviors you describe above, I don't have a preference, but I wouldn't try to do both.

**Comment by The-Compiler** [comment]
Maybe `auto-open` and `auto-open-global` or something?

**Comment by mlochbaum** [comment]
There were a lot of problems. Set `auto-open` to `false` and `show` to `true`, then restart the browser. For me, this broke completions entirely until I changed `auto-open`.

I think I'll just do the by-completion version for now. The other one shouldn't be too hard to implement, but it's not trivial since there isn't any good way right now to figure out when the command ends.
